client/src/main/java/io/pravega/client/stream/impl/ControllerResolverFactory.java

```diff
@@ -137,50 +159,127 @@ public void start(Listener listener) {
                         .usePlaintext(true)
                         .build());
-                this.scheduledExecutor = Executors.newScheduledThreadPool(1,
-                        new ThreadFactoryBuilder().setNameFormat("fetch-controllers-%d").setDaemon(true).build());
-                this.scheduledExecutor.scheduleWithFixedDelay(
-                        this::getControllers, 0L, 120L, TimeUnit.SECONDS);
             } else {
-                // Use the bootstrapped server list as the final set of controllers.
-                final ResolvedServerInfoGroup serverInfoGroup = ResolvedServerInfoGroup.builder().addAll(
-                        this.bootstrapServers.stream().map(ResolvedServerInfo::new)
-                                .collect(Collectors.toList()))
-                        .build();
-                log.info("Updating client with controllers: {}", serverInfoGroup);
-                this.resolverUpdater.onUpdate(Collections.singletonList(serverInfoGroup), Attributes.EMPTY);
+                // If the servers comprise only of IP addresses then we only need to update the controller list once.
+                if (this.bootstrapServers.stream().allMatch(
+                        inetSocketAddress -> InetAddresses.isInetAddress(inetSocketAddress.getHostString()))) {
+                    final ResolvedServerInfoGroup serverInfoGroup = ResolvedServerInfoGroup.builder()
+                            .addAll(this.bootstrapServers.stream()
+                                    .map(address -> new ResolvedServerInfo(
+                                            new InetSocketAddress(address.getHostString(), address.getPort())))
+                                    .collect(Collectors.toList()))
+                            .build();
+                    log.info("Updating client with controllers: {}", serverInfoGroup);
+                    this.resolverUpdater.onUpdate(Collections.singletonList(serverInfoGroup), Attributes.EMPTY);
+                    started.set(true);
+                    return;
+                }
+                // else case is for controllers with DNS hostnames, which needs to be resolved periodically.
             }
+
+            // Schedule the first discovery immediately.
+            this.scheduledExecutor = Executors.newScheduledThreadPool(1,
+                    new ThreadFactoryBuilder().setNameFormat("fetch-controllers-%d").setDaemon(true).build());
+            this.scheduledFuture = this.scheduledExecutor.schedule(this::getControllers, 0L, TimeUnit.SECONDS);
+            started.set(true);
         }
 
+        // This API should not be under a lock as this is already thread safe and we don't want to block the shutdown.
         @Override
-        @Synchronized
         public void shutdown() {
             Preconditions.checkState(started.compareAndSet(true, false));
             if (this.scheduledExecutor != null) {
```

Comment:
A good pattern for this is something along these lines:
```
val e = this.scheduledExecutor;
if (e != null) {
    e.shutdownNow();
    this.scheduledExecutor = null;
}
```
This will clear up the variable once executed. Also, if `this.scheduledExecutor` were an `AtomicReference`, then you could have used the atomic `getAndSet` to ensure that any concurrent calls don't step on each other.